### PR TITLE
[mdatagen] allow filtering out metrics based on resource attributes

### PR DESCRIPTION
**Pull request by povilasv**
**Description:** <Describe what has changed.>

This PR allows filtering out metrics based on resource attributes. For example:

```
        resource_attributes:                                                                      
          k8s.pod.name:
            enabled: true 
            exclude:
              #- strict: "kube-apiserver-kind-control-plane"
              - regexp: "kube-.*"
              - regexp: "coredns-.*"
              - strict: "coredns"
              - strict: "kindnet-mpb2p"
```

Would remove metrics that match regex or strict rules on resource attributes.

**Link to tracking Issue:** <Issue number if applicable>

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25134

**Testing:** <Describe what testing was performed and which tests were added.>
- Tested with k8scluster receiver in kind cluster.
- unit tests added
**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by dmitryax** [comment]
@povilasv can you please rebase? I migrated mdatagen from the contrib repo

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 1]
```diff
@@ -263,6 +283,21 @@ func (mb *MetricsBuilder) EmitForResource(rmo ...ResourceMetricsOption) {
...
+		if filter, ok := mb.resourceAttributeIncludeFilter[name]; ok {
+			if !filter.Matches(val) {
+				return
+			}
+		}
+		if filter, ok := mb.resourceAttributeExcludeFilter[name]; ok {
+			if filter.Matches(val) {
+				return
+			}
+		}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics.go.tmpl:297]: This behavior of combining the `include` and `exclude` looks good to me. I'm curious: what did you use as a reference? Is it aligned with what we have in the filter processor, and is it what the user would expect? This is a pretty important decision and needs to be well documented. I'm mostly talking about the priority choice. 

Also, I think this should also be extracted to the filter module. So, any other implementations get the same behavior.

File: cmd/mdatagen/templates/config.go.tmpl [hunk 2]
```diff
@@ -5,7 +5,9 @@ package {{ .Package }}
 {{ if or .Metrics .ResourceAttributes -}}
 import "go.opentelemetry.io/collector/confmap"
 {{- end }}
-
+{{ if .ResourceAttributes -}}
+import "go.opentelemetry.io/collector/filter"
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply cmd/mdatagen/templates/config.go.tmpl:9]: Nice. Thanks for confirming 👍

File: cmd/mdatagen/templates/testdata/config.yaml.tmpl [hunk 3]
```diff
@@ -29,3 +29,28 @@ none_set:
...
+      enabled: true
+      {{- if eq $attr.Type.String "Str" }}
+      exclude:
+        - strict: "non-existent"
+        - regexp: "non-existent"
+      include:
+        - strict: "non-existent"
+        - regexp: "non-existent"
+        - strict: {{ $attr.TestValue }}
+        - regexp: ".*"
```
Comment (dmitryax) [inline cmd/mdatagen/templates/testdata/config.yaml.tmpl:53]: I think this section should be meaningful like
```suggestion
      exclude:
        - strict: {{ $attr.TestValue }}
      include:
        - regexp: ".*"
```

And in the tests, we validate that no data with string resource attributes is emitted, the same effect as `none_set`

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 4]
```diff
@@ -16,6 +16,9 @@ import (
 	{{- if .SemConvVersion }}
 	conventions "go.opentelemetry.io/collector/semconv/v{{ .SemConvVersion }}"
 	{{- end }}
+	{{ if .ResourceAttributes -}}
+	"go.opentelemetry.io/collector/filter"
```
[the remark this answers is not in the document]
Comment (astencel-sumo) [reply cmd/mdatagen/templates/metrics.go.tmpl:20]: Sounds like it's for defining the configuration, shouldn't it be something like `config/configfilter`?

**Review by dmitryax — APPROVED** [review]

**Pull request merged**